### PR TITLE
Enhance Tests of Source and New Tests of Sink Plugin

### PR DESCRIPTION
**Pull request by poojantcs**

### CODE REVIEW

**Comment by Vipinofficial11** [comment]
@rmstar @itsankit-google Please review.

File: src/e2e-test/features/servicenow/RunTimeWithMacros.feature [hunk 1]
```diff
@@ -84,3 +84,100 @@ Feature: ServiceNow Source - Run time scenarios (macro)
     And Wait till pipeline is in running state
     And Open and capture logs
     Then Verify the pipeline status is "Succeeded"
+
```
Comment (rmstar) [inline src/e2e-test/features/servicenow/RunTimeWithMacros.feature]: let's rename this as we now have sink and multisource tests : servicenow -> servicenowsource

File: src/e2e-test/java/io/cdap/plugin/servicenowsink/locators/ServiceNowSinkPropertiesPage.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.servicenowsink.locators;
+/**
+ * ServiceNow sink - Properties page - Locators.
+ */
+
+public class ServiceNowSinkPropertiesPage {
```
Comment (rmstar) [inline src/e2e-test/java/io/cdap/plugin/servicenowsink/locators/ServiceNowSinkPropertiesPage.java]: unused?

File: src/e2e-test/features/servicenowsink/RunTime.feature [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+    And fill Credentials section for pipeline user
+    And Enter input plugin property: "tableName" with value: "receiving_slip_line"
+    And Select radio button plugin property: "operation" with value: "INSERT"
+    And Validate "ServiceNow" plugin properties
+    And Close the Plugin Properties page
+    And Save and Deploy Pipeline
+    And Run the Pipeline in Runtime
+    And Wait till pipeline is in running state
+    And Verify the pipeline status is "Succeeded"
+    Then Validate OUT record count is equal to IN record count
```
Comment (rmstar) [inline src/e2e-test/features/servicenowsink/RunTime.feature]: for test cases where we run a deployed pipeline, can we also verify that the data written to the sink is correct (not just record count)?

File: src/e2e-test/features/servicenow/RunTimeWithMacros.feature (hunk 1 above)
Comment (Vipinofficial11) [reply src/e2e-test/features/servicenow/RunTimeWithMacros.feature]: Done

File: src/e2e-test/java/io/cdap/plugin/servicenowsink/locators/ServiceNowSinkPropertiesPage.java (hunk 2 above)
Comment (Vipinofficial11) [reply src/e2e-test/java/io/cdap/plugin/servicenowsink/locators/ServiceNowSinkPropertiesPage.java]: yeah it was unused, Deleted it.

File: src/e2e-test/features/servicenowsink/RunTime.feature (hunk 3 above)
Comment (Vipinofficial11) [reply src/e2e-test/features/servicenowsink/RunTime.feature]: Sure, I have created a new Step which fetch response from both the source and sink applications and compare those responses.(commit - bff2432090f29ae80883d1b3539d34f12fe512d0)

File: src/e2e-test/java/io/cdap/plugin/servicenowsink/stepsdesign/DesignTimeSteps.java [hunk 4]
```diff
@@ -16,26 +16,60 @@
...
-  @Then("Verify If new record is created in ServiceNow application")
-  public void verifyIfRecordExistsinServiceNowTable() throws OAuthProblemException, OAuthSystemException {
-    ServiceNowSinkPropertiesPageActions.verifyIfRecordExistsInServiecNowReceivingSlipLineTable(
-      TestSetupHooks.receivingSlipLineRecordUniqueNumber);
- }
+  @Then("Verify If an updated record in ServiceNow application for table {string} is correct")
+  public void verifyIfAnUpdatedRecordInServiceNowApplicationForTableIsCorrect(String tableName)
+    throws OAuthProblemException, OAuthSystemException, IOException, InterruptedException {
+    String tableValueFromPluginPropertiesFile = PluginPropertyUtils.pluginProp(tableName);
+    query = "sys_id=" + TestSetupHooks.systemId;
```
Comment (vikasrathee-cs) [inline src/e2e-test/java/io/cdap/plugin/servicenowsink/stepsdesign/DesignTimeSteps.java]: use ServiceNowConstants variable for sys_id

File: src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java [hunk 5]
```diff
@@ -151,46 +149,51 @@ public String createRecord(String tableName, HttpEntity entity) {
...
+      this.conf.getRestApiEndpoint(), tableName)
+      .setQuery(query);
 
     RestAPIResponse apiResponse = null;
     String accessToken = getAccessToken();
     requestBuilder.setAuthHeader(accessToken);
     apiResponse = executeGet(requestBuilder.build());
+    Gson gson = new Gson();
+    JsonObject jo = gson.fromJson(apiResponse.getResponseBody(), JsonObject.class);
+    JsonArray ja = jo.getAsJsonArray("result");
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java]: give proper names

File: src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java [hunk 6]
```diff
@@ -53,11 +62,93 @@ public static void verifyIfRecordExistsInServiecNowReceivingSlipLineTable(String
...
+    logger.info("Response of ServiceNow : " + responseFromServiceNowTable);
+    Assert.assertTrue(compareValueOfBothResponses(responseFromServiceNowTable, bigQueryResponseInMap));
+  }
+
+  public static void verifyIfRecordUpdatedInServiceNowIsCorrect(String query, String tableName)
+    throws IOException, InterruptedException, OAuthProblemException, OAuthSystemException {
+
+    getRecordFromServiceNowTable(query, tableName);
+    TableResult bigQueryTableData = getBigQueryTableData(TestSetupHooks.bqSourceDataset, TestSetupHooks.bqSourceTable);
+    String bigQueryJsonResponse = bigQueryTableData.getValues().iterator().next().get(0).getValue().toString();
```
Comment (vikasrathee-cs) [inline src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java:90]: this code can throw NullPointerException if table contains no data, you can add a null check to be on the safer side.

File: src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java [hunk 7]
```diff
@@ -53,11 +62,93 @@ public static void verifyIfRecordExistsInServiecNowReceivingSlipLineTable(String
...
+    Gson gson = new Gson();
```
Comment (vikasrathee-cs) [inline src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java]: make it static and at class level and reuse it.

File: src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java [hunk 8]
```diff
@@ -53,11 +62,93 @@ public static void verifyIfRecordExistsInServiecNowReceivingSlipLineTable(String
...
+    JsonObject jsonObject = gson.fromJson(bigQueryJsonResponse, JsonObject.class);
+    Map<String, Object> bigQueryResponseInMap = gson.fromJson(jsonObject.toString(), Map.class);
+
+    logger.info("Response of BigQuery : " + bigQueryResponseInMap);
```
Comment (vikasrathee-cs) [inline src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java]: don't add random loggers

File: src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java [hunk 9]
```diff
@@ -53,11 +62,93 @@ public static void verifyIfRecordExistsInServiecNowReceivingSlipLineTable(String
...
+
+  public static boolean compareValueOfBothResponses(Map<String, String> serviceNowResponseMap,
+                                                    Map<String, Object> bigQueryResponseMap) {
+    if (serviceNowResponseMap.isEmpty() || bigQueryResponseMap.isEmpty()) {
+      return false;
+    }
+    boolean result = false;
+    Set<String> bigQueryKeySet = bigQueryResponseMap.keySet();
+
+    for (String setValue : bigQueryKeySet) {
```
Comment (vikasrathee-cs) [inline src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java]: use key instead of setValue

File: src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java [hunk 10]
```diff
@@ -53,11 +62,93 @@ public static void verifyIfRecordExistsInServiecNowReceivingSlipLineTable(String
...
+        String bigQueryFormattedValue = serviceNowDateFormat.format(checkBigQueryDateFormat(bigQueryValue.toString()));
+        result = String.valueOf(serviceNowValue).equals(bigQueryFormattedValue);
+      } else {
+        result = String.valueOf(serviceNowValue).equals(String.valueOf(bigQueryValue));
+      }
+    }
+    return result;
+  }
 
+  public static Date checkBigQueryDateFormat(String value) {
```
Comment (vikasrathee-cs) [inline src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java]: make this method private and add description on top

File: src/e2e-test/java/io/cdap/plugin/servicenowsink/stepsdesign/DesignTimeSteps.java (hunk 4 above)
Comment (Vipinofficial11) [reply src/e2e-test/java/io/cdap/plugin/servicenowsink/stepsdesign/DesignTimeSteps.java]: Done

File: src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java (hunk 5 above)
Comment (Vipinofficial11) [reply src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java]: Changed the names

File: src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java (hunk 6 above)
Comment (Vipinofficial11) [reply src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java:90]: Added a check.

File: src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java (hunk 7 above)
Comment (Vipinofficial11) [reply src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java]: Done.

File: src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java (hunk 8 above)
Comment (Vipinofficial11) [reply src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java]: Removed the loggers.

File: src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java (hunk 9 above)
Comment (Vipinofficial11) [reply src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java]: changed: setValues -> key

File: src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java (hunk 10 above)
Comment (Vipinofficial11) [reply src/e2e-test/java/io/cdap/plugin/servicenowsink/actions/ServiceNowSinkPropertiesPageActions.java]: Done

File: .github/workflows/e2e.yml [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#  License for the specific language governing permissions and limitations under
+#  the License.
+
+# This workflow will build a Java project with Maven
+# For more information see: https://help.github.com/actions/language-and-framework-guides/building-and-testing-java-with-maven
+# Note: Any changes to this workflow would be used only after merging into develop
+name: Build e2e tests
```
Comment (itsankit-google) [inline .github/workflows/e2e.yml:15]: Can the setup files be merged first in a separate PR?

**Comment by itsankit-google** [comment]
Please resolve conflicts.

File: .github/workflows/e2e.yml (hunk 11 above)
Comment (Vipinofficial11) [reply .github/workflows/e2e.yml:15]: Sure Ankit I am raising Two different Pr's and closing this one.
